elastic/beats
# Move `umask` from code to service files

Enhancement

## What does this PR do?



    
Before this the `umask` value was hard-coded in `libbeat`. It caused
some confusion among the users since file permission configuration was
practically ignored by a beat on the level of the binary. It's been
decided that we move `umask` to the service files, so the distribution
is secured by default but it still allows the users to set the value
if they choose to.

## Why is it important?


The issue started as a security concern here https://github.com/elastic/beats/issues/14005

Then was addressed by adding a hard-coded umask for all files created by `libbeat` https://github.com/elastic/beats/pull/14119

Later documented here https://github.com/elastic/beats/pull/28347

We've got some feedback from our customers about confusion and lack of documentation. Examples can be found here https://github.com/elastic/beats/issues/20584#issuecomment-905014677

## Checklist



~~- [ ] I have commented my code, particularly in hard-to-understand areas~~
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist


- [x] systemd unit sets umask to 0027 for both beats and elastic agent
- [x] launchd manifest contains umask property set to 0027 for both beats and elstic agent
- [x] shell scripts that start both elastic agent and beats contain `umask 0027` command

I made sure that all current permission defaults in beats are secure (set to `0600`):

libbeat

* Output files
  * Config https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/outputs/fileout/config.go#L41 
  * Reference https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/_meta/config/output-file.reference.yml.tmpl#L33 
  * Documentation https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/outputs/fileout/docs/fileout.asciidoc#permissions

* Log files
  * Config https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/logp/config.go#L78
  * Reference https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/_meta/config/logging.reference.yml.tmpl#L58
  * Documentation  https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/libbeat/docs/loggingconfig.asciidoc#loggingfilespermissions

filebeat

* Registry
  * Config https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/filebeat/config/config.go#L65
  * Reference https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/filebeat/_meta/config/filebeat.global.reference.yml.tmpl#L10
  * Documentation https://github.com/elastic/beats/blob/cbdba8aa1645a896023e387b370b5b88960f3176/filebeat/docs/filebeat-general-options.asciidoc#registryfile_permissions

## How to test this PR locally


I configured filebeat to have a file output with 666 permissions. 

For that one can use the following config section in the filebeat config file:

```yaml
output.file:
   path: "/path/to/the/output/directory"
   filename: "output.log"
   permissions: 0666
```

Before this change this would results in a file with `640` permissions, after this change it's `644` which lets users to make files available to the world as they request.

After setting `umask 027` prior to running the filebeat binary the output file is created with `640` as expected.

## Related issues


- Closes #25455
